core/java/com/google/instrumentation/common/MatchFunction.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright 2016, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.common;
+
+/**
+ * Encodes functions {@code func: A -> B} for use in encoding disjoint-unions.
```

Comment:
I've updated the description slightly and added a link to AggregationDescriptor and will consider how to show a simple example.

I've put this class in common because several of our datastructures will need to encode tagged unions (aka sum types, discriminated unions ...) in Java.

I've considered a number of approaches (matching functions, the visitor pattern, istanceof, and so on) and I find this to be the most clear and type safe - I am open to alternatives if there are compelling arguments.